src/gen-mapping.ts

```diff
@@ -336,6 +350,98 @@ export class GenMapping {
           : [genColumn, sourcesIndex, sourceLine, sourceColumn],
       );
     };
+
+    applySourceMap = (
+      map: GenMapping,
+      consumer: TraceMap,
+      rawSourceFile?: string,
+      sourceMapPath?: string,
+    ) => {
+      let sourceFile = rawSourceFile;
+
+      if (sourceFile == null) {
+        if (consumer.file == null) {
+          throw new Error(
+            'applySourceMap requires either an explicit source file, ' +
+              'or the source map\'s "file" property. Both were omitted.',
+          );
+        }
+        sourceFile = consumer.file;
+      }
+
+      const sourceRoot = map.sourceRoot;
+
+      // Make "sourceFile" relative if an absolute Url is passed.
+      if (sourceRoot != null) {
+        sourceFile = relative(sourceRoot, sourceFile);
+      }
+
+      const sourceIndex = get(map._sources, sourceFile);
+
+      // If the applied source map replaces a source entirely
+      // it's better to start with a fresh set of source, sourceContent and names
+      const newSources = new SetArray();
+      const newSourceContents: (string | null)[] = [];
+      const newNames = new SetArray();
+
+      for (const line of map._mappings) {
+        for (const seg of line) {
+          if (seg.length === 1) continue;
+
+          if (seg[1] !== sourceIndex) {
+            // This isn't the file we want to remap; keep the original mapping
+            const initialIndex = seg[1];
+            seg[1] = put(newSources, map._sources.array[initialIndex]);
+            newSourceContents[seg[1]] = map._sourcesContent[initialIndex];
+            if (seg.length === 5) {
+              seg[4] = put(newNames, map._names.array[seg[4]]);
+            }
+            continue;
+          }
+
+          const traced = traceSegment(consumer, seg[2], seg[3]);
+          if (traced == null) {
+            // Could not find a mapping; keep the original mapping
+            const initialIndex = seg[1];
+            seg[1] = put(newSources, map._sources.array[initialIndex]);
+            newSourceContents[seg[1]] = map._sourcesContent[initialIndex];
+            if (seg.length === 5) {
+              seg[4] = put(newNames, map._names.array[seg[4]]);
+            }
+            continue;
+          }
+
+          let source = consumer.sources[traced[1] as number] as string;
```

Comment:
This is a different segment. `seg` has been checked, meaning the segment in our map has a `sourceIndex`/`sourceLine`/`sourceColumn`. We've then traced that location through the consumer, which returns a new segment `traced`. `traced` can be `1` length, meaning it doesn't correspond to a source location in _that_ map.

Reply:
oh right. but in that case wouldn't it implicit that `traceSegment` can't return a element that's 1 in length ? Since that would not be a resolved path.

Sorry if I'm misunderstanding sourcemaps, It's the first time I'm really digging in the topic.